README.md

```diff
@@ -1,2 +1,200 @@
+but these don't always do the job, just because NAT traversal is complicatd.
+That's why it's useful to have a simple relay protocol.
+
+One node asks a relay node to connect to another node on behalf of the first node.
+The relay node shortcircuits its streams to the two nodes,
+and they are then connected through the relay.
+This connection behaves just like a direct connection would,
+because it is in fact a direct connection,
+but over an existing swarm stream with another peer, instead of e.g. TCP.
+
+Relayed connections are end-to-end encrypted just like direct connections.
+
+The circuit relay is both a tunneled transport and a mounted swarm protocol.
+The transport is the means to *establishing* and *accepting* connections,
+and the swarm protocol is the means to *relaying* connections.
+
+```
++-------+    /ip4/.../tcp/.../ws/ipfs/QmRelay    +---------+    /ip4/.../tcp/.../ipfs/QmTwo        +-------+
+| QmOne | <------------------------------------> | QmRelay | <-----------------------------------> | QmTwo |
++-------+    (/ipfs/relay/circuit multistream)   +---------+   (/ipfs/relay/circuit multistream)   +-------+
+      ^                                            +-----+                                           ^
+      |                                            |     |                                           |
+      |            /p2p-circuit/QmTwo              |     |                                           |
+      +--------------------------------------------+     +-------------------------------------------+
+```
+
+TODO: the stream codec should be called `/libp2p/relay/circuit` instead.
+
+Note: we're using the `/p2p` multiaddr protocol instead of `/ipfs` in this document.
+`/ipfs` is currently the canonical way of addressing a libp2p or IPFS node,
+but given the growing non-IPFS usage of libp2p, we'll migrate to using `/p2p`.
+
+Note: at the moment we're not including a mechanism for discovering relay nodes.
+For the time being, they should be configured statically.
+
+
+## Dramatization
+
+Cast:
+- QmOne, the dialing node (browser).
+- QmTwo, the listening node (go-ipfs).
+- QmRelay, a node which speaks the circuit relay protocol (go-ipfs or js-ipfs).
+
+Scene 1:
+- QmOne wants to connect to QmTwo,
+  and through peer routing has acquired a set of addresses of QmTwo.
+- QmTwo doesn't support any of the transports used by QmOne.
+- Awkward silence.
+
+Scene 2:
+- All three nodes have learned to speak the `/ipfs/relay/circuit` protocol.
+- QmRelay is configured to allow relaying connections between other nodes.
+- QmOne is configured to use QmRelay for relaying.
+- QmOne automatically added `/p2p-circuit/ipfs/QmTwo` to its set of QmTwo addresses.
+- QmOne tries to connect via relaying, because it shares this transport with QmTwo.
+- A lively and prolonged dialogue ensues.
+
+
+## Addressing
+
+`/p2p-circuit` multiaddrs don't carry any meaning of their own.
```

Comment:
I'm not sure how to avoid that:

1. multiaddrs are meant to be as verbose as neccessary.
2. some kind of prefix for libp2p-specific protocols (other example: `/libp2p-webrtc-star`) is likely helpful to open up development to other projects and communities.

I also don't mind it much really :)